public/scripts/sentenceBuilder.js

```diff
@@ -0,0 +1,359 @@
+  if (REQUIRED_KEYWORD_COUNT) {
+    updateChecklistCounter(requiredWordCounter, currentRequiredWordCount, REQUIRED_KEYWORD_COUNT, false)
+  } else {
+    requiredWordCounter.textContent = ' N/A'
+  }
+  const errorCount = Object.keys(oldClasses).reduce((count, index) => {
+    if (oldClasses[index] === 'error' || oldClasses[index] === 'duplicate') {
+      ++count
+    }
+    return count
+  }, 0)
+  updateChecklistCounter(errorCounter, errorCount, 0)
+}
+
+function updateChecklistCounter(counterElem, curCount, requiredCount, needsExactCount = true) {
+  const firstPart = `${curCount < 10 ? '&nbsp;' : ''}${curCount}`
+  const secondPart = requiredCount ? `/${requiredCount}` : ''
+  counterElem.innerHTML = `${firstPart}${secondPart}`
+  if (needsExactCount ? curCount !== requiredCount : curCount < requiredCount) {
+    counterElem.classList.add('error')
+  } else {
+    counterElem.classList.remove('error')
+  }
+}
+
+function moveCursorLeft({
+  ctrlKey,
+  shiftKey
+} = {}) {
+  if (ctrlKey) {
+    do {
+      moveCursorLeft()
+    } while (cursorPos !== 0 && sentence.charAt(cursorPos - 1) !== ' ')
+  } else if (shiftKey) {
+    cursorPos = 0
+  } else {
+    cursorPos && --cursorPos
+  }
+}
+
+function moveCursorRight({
+  ctrlKey,
+  shiftKey
+} = {}) {
+  if (ctrlKey) {
+    while (cursorPos !== sentence.length && sentence.charAt(cursorPos) !== ' ') {
+      moveCursorRight()
+    }
+    moveCursorRight()
+  } else if (shiftKey) {
+    cursorPos = sentence.length
+  } else {
+    cursorPos < sentence.length && ++cursorPos
+  }
+}
+
+function getWordIndexOfCursor(wordsArray) {
+  let letterCount = 0
+
+  for (let i = 0; i !== wordsArray.length; ++i) {
+    const wordLength = wordsArray[i].length
+    letterCount += wordLength
+    if (letterCount >= cursorPos) {
+      return {
+        word: i,
+        letter: cursorPos - letterCount + wordLength
+      }
+    }
+    ++letterCount
+  }
+
+  return {
+    word: 0,
+    letter: 0
+  }
+}
+
+function getSpanHTML(word, wordId, className = oldClasses[wordId]) {
+  return `<span id="${wordId}" ${className ? `class="${className}"` : ''}>${word}</span>`
+}
+
+function classifyWords(wordsArray, randomWords) {
+  wordsArray = wordsArray.filter(word => word !== '')
+  const wordClasses = []
+  const indexesToCheck = []
+  const wordSet = {}
+
+  clearMarkedRequiredWords()
+  wordsArray.forEach((word, i) => {
+    if (wordSet[word] !== undefined) {
+      wordSet[word].push(i)
+      wordSet[word].forEach(index => {
+        indexesToCheck.splice(indexesToCheck.indexOf(index), 1)
+        wordClasses[index] = 'duplicate'
+      })
+    } else {
+      wordSet[word] = [i]
+      if (RANDOM_WORDS.includes(word)) {
+        // words.querySelector(`#word${i}`).className = 'keyword'
+        wordClasses[i] = 'keyword'
+        markRequiredWord(word)
+      } else {
+        indexesToCheck.push(i)
+      }
+    }
+  })
+  const wordsToCheck = wordsArray.filter((_, i) => indexesToCheck.includes(i))
```

Comment:
`_` here is a convention where you're saying "I don't really care about that param. It's not special syntax or anything, `_` is just a valid identifier that look kind of cool as a throwaway name